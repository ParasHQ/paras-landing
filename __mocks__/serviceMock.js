import axios from 'axios'

const BASE_API = process.env.V2_API_URL

export const fetchTokenTest = async () => {
	try {
		const resp = await axios.get(`${BASE_API}/token`, {
			params: {
				contract_id: 'x.paras.near',
				token_series_id: '230098',
				token_id: '230098:1',
				lookup_token: true,
			},
		})
		const token = resp.data.data.results[0]

		const url = resp.config.url
		const contractId = token.contract_id
		const tokenId = token.token_id
		const ftTokenId = token.ft_token_id
		const price = token.price
		const hasPrice = token.has_price

		if (!token.has_price) return 'This token is not for sale.'

		return {
			url,
			contract_id: contractId,
			token_id: tokenId,
			ft_token_id: ftTokenId,
			price,
			has_price: hasPrice,
			result: token,
		}
	} catch (err) {
		// eslint-disable-next-line no-console
		console.log('error message: ', err)
		return []
	}
}
