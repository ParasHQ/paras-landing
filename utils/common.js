import CID from 'cids'
import Compressor from 'compressorjs'
import TimeAgo from 'javascript-time-ago'
import en from 'javascript-time-ago/locale/en'
import crypto from 'crypto'

TimeAgo.addLocale(en)
export const timeAgo = new TimeAgo('en-US')

const monthNames = [
	'January',
	'February',
	'March',
	'April',
	'May',
	'June',
	'July',
	'August',
	'September',
	'October',
	'November',
	'December',
]

export const parseDate = (ts) => {
	let dateObj = new Date(ts)
	let month = monthNames[dateObj.getMonth()].slice(0, 3)
	let day = String(dateObj.getDate()).padStart(2, '0')
	let year = dateObj.getFullYear()
	return `${day} ${month} ${year}`
}

export const prettyBalance = (balance, decimals = 18, len = 8) => {
	if (!balance) {
		return '0'
	}
	const diff = balance.toString().length - decimals
	const fixedPoint = Math.max(2, len - Math.max(diff, 0))
	const fixedBalance = (balance / 10 ** decimals).toFixed(fixedPoint)
	const finalBalance = parseFloat(fixedBalance).toString()
	const [head, tail] = finalBalance.split('.')
	if (head == 0) {
		if (tail) {
			return `${head}.${tail.substring(0, len - 1)}`
		}
		return `${head}`
	}
	const formattedHead = head.replace(/\B(?=(\d{3})+(?!\d))/g, ',')
	return tail ? `${formattedHead}.${tail}` : formattedHead
}

export const prettyTruncate = (str = '', len = 8, type) => {
	if (str && str.length > len) {
		if (type === 'address') {
			const front = Math.ceil(len / 2)
			const back = str.length - (len - front)
			return `${str.slice(0, front)}...${str.slice(back)}`
		}
		return `${str.slice(0, len)}...`
	}
	return str
}

export const readFileAsUrl = (file) => {
	const temporaryFileReader = new FileReader()

	return new Promise((resolve) => {
		temporaryFileReader.onload = () => {
			resolve(temporaryFileReader.result)
		}
		temporaryFileReader.readAsDataURL(file)
	})
}

export const readFileDimension = (file) => {
	const temporaryFileReader = new FileReader()

	return new Promise((resolve) => {
		temporaryFileReader.onload = () => {
			const img = new Image()

			img.onload = () => {
				resolve({
					width: img.width,
					height: img.height,
				})
			}

			img.src = temporaryFileReader.result
		}
		temporaryFileReader.readAsDataURL(file)
	})
}

export const parseImgUrl = (url, defaultValue = '', opts = {}) => {
	if (!url) {
		return defaultValue
	}
	if (url.includes('://')) {
		const [protocol, path] = url.split('://')
		if (protocol === 'ipfs') {
			const cid = new CID(path)
			if (opts.useOriginal || process.env.APP_ENV !== 'production') {
				if (cid.version === 0) {
					return `https://ipfs-gateway.paras.id/ipfs/${path}`
				} else {
					return `https://ipfs.fleek.co/ipfs/${path}`
				}
			}

			let transformationList = []
			if (opts.width) {
				transformationList.push(`w=${opts.width}`)
				transformationList.push(`auto=format,compress`)
			} else {
				transformationList.push('w=800')
				transformationList.push(`auto=format,compress`)
			}
			return `https://paras-cdn.imgix.net/${cid}?${transformationList.join('&')}`
		} else if (opts.isMediaCdn) {
			const sha1Url = sha1(url)
			let transformationList = []
			if (opts.width) {
				transformationList.push(`w=${opts.width}`)
				transformationList.push(`auto=format,compress`)
			} else {
				transformationList.push('w=800')
				transformationList.push(`auto=format,compress`)
			}
			return `https://paras-cdn.imgix.net/${sha1Url}?${transformationList.join('&')}`
		}
		return url
	} else {
		try {
			const cid = new CID(url)
			if (opts.useOriginal || process.env.APP_ENV !== 'production') {
				if (cid.version === 0) {
					return `https://ipfs-gateway.paras.id/ipfs/${cid}`
				} else if (cid.version === 1) {
					return `https://ipfs.fleek.co/ipfs/${cid}`
				}
			}

			let transformationList = []
			if (opts.width) {
				transformationList.push(`w=${opts.width}`)
				transformationList.push(`auto=format,compress`)
			} else {
				transformationList.push('w=800')
				transformationList.push(`auto=format,compress`)
			}
			return `https://paras-cdn.imgix.net/${cid}?${transformationList.join('&')}`
		} catch (err) {
			return url
		}
	}
}

export const dataURLtoFile = (dataurl, filename) => {
	let arr = dataurl.split(','),
		mime = arr[0].match(/:(.*?);/)[1],
		bstr = atob(arr[1]),
		n = bstr.length,
		u8arr = new Uint8Array(n)

	while (n--) {
		u8arr[n] = bstr.charCodeAt(n)
	}

	return new File([u8arr], filename, { type: mime })
}

export const compressImg = (file) => {
	return new Promise((resolve, reject) => {
		let _file = file
		const quality = 0.8
		new Compressor(_file, {
			quality: quality,
			maxWidth: 1080,
			maxHeight: 1080,
			convertSize: Infinity,
			success: resolve,
			error: reject,
		})
	})
}

export const checkUrl = (str) => {
	var pattern = new RegExp(
		'^(https?:\\/\\/)?' +
			'((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' +
			'((\\d{1,3}\\.){3}\\d{1,3}))' +
			'(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' +
			'(\\?[;&a-z\\d%_.~+=-]*)?' +
			'(\\#[-a-z\\d_]*)?$',
		'i'
	)
	return !!pattern.test(str)
}

export const parseSortQuery = (sort, defaultMinPrice = false) => {
	if (!sort) {
		return defaultMinPrice ? 'lowest_price::1' : 'updated_at::-1'
	} else if (sort === 'marketupdate') {
		return 'updated_at::-1'
	} else if (sort === 'marketupdateasc') {
		return 'updated_at::1'
	} else if (sort === 'cardcreate') {
		return '_id::-1'
	} else if (sort === 'cardcreateasc') {
		return '_id::1'
	} else if (sort === 'pricedesc') {
		return 'lowest_price::-1'
	} else if (sort === 'priceasc') {
		return 'lowest_price::1'
	}
}

export const parseSortTokenQuery = (sort) => {
	if (!sort || sort === 'cardcreate') {
		return '_id::-1'
	} else if (sort === 'cardcreateasc') {
		return '_id::1'
	} else if (sort === 'pricedesc') {
		return 'price::-1'
	} else if (sort === 'priceasc') {
		return 'price::1'
	}
}

export const parseGetTokenIdfromUrl = (url) => {
	const pathname = new URL(url).pathname.split('/')
	return {
		token_series_id: pathname[2],
		token_id: pathname[3],
	}
}

export const parseGetCollectionIdfromUrl = (url) => {
	const pathname = new URL(url).pathname.split('/')
	return {
		collection_id: pathname[2],
	}
}

export const capitalize = (words) => {
	return words[0].toUpperCase() + words.slice(1)
}

export default function sha1(data, encoding) {
	return crypto
		.createHash('sha1')
		.update(data)
		.digest(encoding || 'hex')
}
