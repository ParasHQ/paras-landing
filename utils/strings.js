import LocalizedStrings from 'react-localization'
import { defaultsDeep } from 'lodash'
import english from '../public/static/labels/english'

export const locales = {
	en: {
		displayName: 'English',
		localeCode: 'en',
		labels: english,
	},
	cn: {
		displayName: 'Chinese',
		localeCode: 'cn',
		labels: () => import(`public/static/labels/chinese`),
	},
}

export const strings =
	new LocalizedStrings({
		default: locales.en.labels,
	}) || english

export const getLocaleStrings = async ({ locale }) => {
	if (locale === 'en' || !(locale in locales)) {
		return english
	}
	const localeStrings = locales[locale].labels().then((l) => l.default)
	return defaultsDeep(localeStrings, english)
}
