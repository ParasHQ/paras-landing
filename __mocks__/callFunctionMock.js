import { GAS_FEE, GAS_FEE_150 } from 'config/constants'
import { fetchTokenTest } from './serviceMock'

const CONTRACT_ID = process.env.MARKETPLACE_CONTRACT_ID

export const onBuyToken = async () => {
	const params = await fetchTokenTest()

	return {
		contractId: CONTRACT_ID,
		methodName: `buy`,
		args: {
			token_id: params.token_id,
			nft_contract_id: params.contract_id,
			ft_token_id: params.ft_token_id,
			price: params.price,
		},
		gas: GAS_FEE_150,
		deposit: params.price,
	}
}

export const onPlaceBid = async (value) => {
	const params = await fetchTokenTest()

	return {
		receiverId: CONTRACT_ID,
		methodName: 'add_offer',
		args: {
			nft_contract_id: params.contract_id,
			token_id: params.token_id,
			ft_token_id: params.ft_token_id,
			price: value,
		},
		deposit: value,
		gas: GAS_FEE,
	}
}
