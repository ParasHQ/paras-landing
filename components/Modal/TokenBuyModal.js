import { useEffect, useState } from 'react'
import Button from 'components/Common/Button'
import Modal from 'components/Common/Modal'
import { formatNearAmount } from 'near-api-js/lib/utils/format'
import LoginModal from './LoginModal'
import { GAS_FEE_150 } from 'config/constants'
import { IconX } from 'components/Icons'
import { useIntl } from 'hooks/useIntl'
import { sentryCaptureException } from 'lib/sentry'
import { trackBuyToken, trackBuyTokenImpression } from 'lib/ga'
import useProfileData from 'hooks/useProfileData'
import { flagColor, flagText } from 'constants/flag'
import BannedConfirmModal from './BannedConfirmModal'
import useStore from 'lib/store'
import WalletHelper from 'lib/WalletHelper'

const TokenBuyModal = ({ show, onClose, data }) => {
	const [showLogin, setShowLogin] = useState(false)
	const { currentUser } = useStore()
	const [showBannedConfirm, setShowBannedConfirm] = useState(false)
	const creatorData = useProfileData(data.metadata.creator_id)

	const { localeLn } = useIntl()

	useEffect(() => {
		if (show) {
			trackBuyTokenImpression(data.token_id)
		}
	}, [show])

	const onBuyToken = async () => {
		if (!currentUser) {
			setShowLogin(true)
			return
		}

		trackBuyToken(data.token_id)

		try {
			const params = {
				token_id: data.token_id,
				nft_contract_id: data.contract_id,
				ft_token_id: data.ft_token_id,
				price: data.price,
			}

			await WalletHelper.callFunction({
				contractId: process.env.MARKETPLACE_CONTRACT_ID,
				methodName: `buy`,
				args: params,
				gas: GAS_FEE_150,
				deposit: data.price,
			})

			// TODO After Function Call Sender Wallet
		} catch (err) {
			sentryCaptureException(err)
		}
	}

	return (
		<>
			<Modal isShow={show} closeOnBgClick={false} closeOnEscape={false} close={onClose}>
				<div className="max-w-sm w-full p-4 bg-gray-800 m-4 md:m-auto rounded-md relative">
					<div className="absolute right-0 top-0 pr-4 pt-4">
						<div className="cursor-pointer" onClick={onClose}>
							<IconX />
						</div>
					</div>
					<div>
						<h1 className="text-2xl font-bold text-white tracking-tight">
							{localeLn('ConfirmBuy')}
						</h1>
						<p className="text-white mt-2">
							{localeLn('AreAboutToPurchase')} <b>{data.metadata.title}</b>
						</p>
						<div className="mt-4">
							<div className="mt-2 text-sm text-red-500"></div>
						</div>
						<div className="mt-4 text-center">
							<div className="text-white my-1">
								<div className="flex justify-between">
									<div className="text-sm">{localeLn('Total')}</div>
									<div className="text">{data.price && `${formatNearAmount(data.price)} â“ƒ`}</div>
								</div>
							</div>
						</div>
						{creatorData?.flag && (
							<div className="z-20 bottom-0 flex items-center justify-center px-4 mt-4 w-full">
								<p
									className={`text-white text-sm m-2 mt-2 p-1 font-bold w-full mx-auto px-4 text-center rounded-md ${
										flagColor[creatorData?.flag]
									}`}
								>
									{localeLn(flagText[creatorData?.flag])}
								</p>
							</div>
						)}
						<p className="text-white mt-4 text-sm text-center opacity-90">
							{localeLn('RedirectedToconfirm')}
						</p>
						<div className="mt-6">
							<Button
								size="md"
								isFullWidth
								onClick={() => (creatorData?.flag ? setShowBannedConfirm(true) : onBuyToken())}
							>
								{data.price !== '0' ? localeLn('Buy') : localeLn('GetForFree')}
							</Button>
							<Button variant="ghost" size="md" isFullWidth className="mt-4" onClick={onClose}>
								{localeLn('Cancel')}
							</Button>
						</div>
					</div>
				</div>
			</Modal>
			{showBannedConfirm && (
				<BannedConfirmModal
					creatorData={creatorData}
					action={onBuyToken}
					setIsShow={(e) => setShowBannedConfirm(e)}
					onClose={onClose}
				/>
			)}
			<LoginModal onClose={() => setShowLogin(false)} show={showLogin} />
		</>
	)
}

export default TokenBuyModal
