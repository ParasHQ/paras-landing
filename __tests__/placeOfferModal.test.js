import '@testing-library/jest-dom'
import { fireEvent, render, screen } from '@testing-library/react'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import TokenBuyModal from 'components/Modal/TokenBuyModal'
import { IntlProvider } from 'react-intl'
import { onPlaceBid } from '__mocks__/callFunctionMock'
import { fetchTokenTest } from '__mocks__/serviceMock'
import esTranslations from '__mocks__/locale/en.json'
import { parseNearAmount } from 'near-api-js/lib/utils/format'
import PlaceOfferModal from 'components/Modal/PlaceOfferModal'

const tokens = {
	contract_id: 'x.paras.near',
	token_series_id: '230098',
	token_id: '230098:1',
	ft_token_id: 'near',
	price: parseNearAmount('5.7'),
	has_price: true,
}

const valueOffer = 0.1

describe('PlaceOfferModal Testing', () => {
	let mock

	beforeAll(async () => {
		mock = new MockAdapter(axios)
		window.gtag = function () {}
	})

	afterEach(() => {
		mock.reset()
	})

	describe('API Gateway', () => {
		describe('When API call is successful', () => {
			test('should return contract_id && token_id', async () => {
				// given
				mock.onGet(`${process.env.V2_API_URL}/token`).reply(200, tokens)

				mock.restore()

				// // when
				const resultMock = mock.handlers.get[0]
				const resultAPI = await fetchTokenTest()

				// then
				expect(resultMock[0]).toEqual(resultAPI.url)
				expect(resultMock[4].contract_id).toEqual(resultAPI.contract_id)
				expect(resultMock[4].token_id).toEqual(resultAPI.token_id)
				expect(resultMock).toMatchSnapshot()
				expect(resultAPI).toMatchSnapshot()
			})
		})

		describe('When API call fails', () => {
			test('should return empty data', async () => {
				mock.onGet(`${process.env.V2_API_URL}/token`).networkErrorOnce()

				const resultMock = mock.handlers.get[0]
				const resultAPI = await fetchTokenTest()

				expect(resultMock[0]).toEqual(resultAPI.url)
				expect(resultMock[4] === undefined && []).toEqual([])
			})
		})
	})

	describe('callFunction => method: add_offer', () => {
		test('should input offer greater than 0.0', async () => {
			const token = await fetchTokenTest()
			render(
				<IntlProvider locale="en" messages={esTranslations}>
					<PlaceOfferModal data={token.result} show />
				</IntlProvider>
			)

			const inputOffer = screen.getByPlaceholderText('Place your Offer')
			fireEvent.change(inputOffer, { target: { value: 0.0 } })

			expect(valueOffer).toBeGreaterThan(Number(inputOffer.value))
			expect(inputOffer.value).toMatchSnapshot()
		})

		test('should appear input offer', async () => {
			const token = await fetchTokenTest()
			render(
				<IntlProvider locale="en" messages={esTranslations}>
					<PlaceOfferModal data={token.result} show />
				</IntlProvider>
			)

			const inputOffer = fireEvent.click(screen.getByRole('spinbutton'))

			expect(inputOffer).toBeTruthy()
			expect(inputOffer).toMatchSnapshot()
		})

		test('should return arguments and method the same like data API', async () => {
			mock.onGet(`${process.env.V2_API_URL}/token`).reply(200, tokens)

			mock.restore()

			const tokenMock = mock.handlers.get[0][4]
			const token = await fetchTokenTest()
			const callFunction = await onPlaceBid(valueOffer)

			const callFunctionMock = jest.fn()
			callFunctionMock.mockReturnValueOnce({
				receiverId: process.env.MARKETPLACE_CONTRACT_ID,
				methodName: 'add_offer',
				args: {
					nft_contract_id: tokenMock.contract_id,
					token_id: tokenMock.token_id,
					ft_token_id: tokenMock.ft_token_id,
					price: valueOffer,
				},
				deposit: valueOffer,
				gas: `100000000000000`,
			})

			render(
				<IntlProvider locale="en" messages={esTranslations}>
					<TokenBuyModal data={token.result} show />
				</IntlProvider>
			)

			expect(callFunctionMock()).toMatchObject(callFunction)
			expect(callFunctionMock).toMatchSnapshot()
		})
	})
})
