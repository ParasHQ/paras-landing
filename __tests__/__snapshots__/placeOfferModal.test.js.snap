// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceOfferModal Testing API Gateway When API call is successful should return contract_id && token_id 1`] = `
Array [
  "https://api-v2-mainnet.paras.id/token",
  undefined,
  undefined,
  200,
  Object {
    "contract_id": "x.paras.near",
    "ft_token_id": "near",
    "has_price": true,
    "price": "5700000000000000000000000",
    "token_id": "230098:1",
    "token_series_id": "230098",
  },
  undefined,
]
`;

exports[`PlaceOfferModal Testing API Gateway When API call is successful should return contract_id && token_id 2`] = `
Object {
  "contract_id": "x.paras.near",
  "ft_token_id": "near",
  "has_price": true,
  "price": "5700000000000000000000000",
  "result": Object {
    "_id": "620f98bf089d2c5bb02c8d93",
    "approval_id": 9,
    "categories": Array [],
    "contract_id": "x.paras.near",
    "edition_id": "1",
    "ft_token_id": "near",
    "has_price": true,
    "has_rank": true,
    "is_creator": true,
    "likes": null,
    "metadata": Object {
      "blurhash": "U01CvZMwo#fPj@tSf6j]o#f6j[aykDj[j[j[",
      "collection": "Soul Out",
      "collection_id": "soul-out-by-sxthyboynear",
      "copies": 1,
      "creator_id": "sxthyboy.near",
      "description": "\\" โฮกปิ้บ \\"
\\" Soul Out \\"collection is 111 unique NFT in NEAR Protocol Blockchain Only. It is all hand-drawn art.",
      "expires_at": null,
      "extra": null,
      "issued_at": "1645189307622433324",
      "media": "bafybeihbhlksad3d336hxuayqoosplfxe2tw5vw43fxevu4ta7kagj42f4",
      "media_hash": null,
      "mime_type": "image/png",
      "rank": 35,
      "reference": "bafkreidv6pipty4oknf5xo4zqusdgytrblzwgpw7oualwudfmhp67sa4dq",
      "reference_hash": null,
      "score": 0,
      "starts_at": null,
      "title": "Soul Out #024 #1",
      "updated_at": null,
    },
    "owner_id": "iqbalutomo.near",
    "price": "5700000000000000000000000",
    "royalty": Object {
      "sxthyboy.near": 1000,
    },
    "token_id": "230098:1",
    "token_series_id": "230098",
    "view": 111,
  },
  "token_id": "230098:1",
  "url": "https://api-v2-mainnet.paras.id/token",
}
`;

exports[`PlaceOfferModal Testing callFunction => method: add_offer should appear input offer 1`] = `true`;

exports[`PlaceOfferModal Testing callFunction => method: add_offer should input offer greater than 0.0 1`] = `"0"`;

exports[`PlaceOfferModal Testing callFunction => method: add_offer should return arguments and method the same like data API 1`] = `
[MockFunction] {
  "calls": Array [
    Array [],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "args": Object {
          "ft_token_id": "near",
          "nft_contract_id": "x.paras.near",
          "price": 0.1,
          "token_id": "230098:1",
        },
        "deposit": 0.1,
        "gas": "100000000000000",
        "methodName": "add_offer",
        "receiverId": "marketplace.paras.near",
      },
    },
  ],
}
`;
